lib/datadog/core/environment/git.rb

```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+require_relative '../git/ext'
+require_relative '../utils/url'
+
+module Datadog
+  module Core
+    module Environment
+      # Retrieves git repository information from environment variables
+      module Git
+        module_function
+
+        def git_repository_url
+          return @git_repository_url if defined?(@git_repository_url)
+
+          @git_repository_url = Utils::Url.filter_basic_auth(ENV[Datadog::Core::Git::Ext::ENV_REPOSITORY_URL])
```

Comment:
I suspect this can be a bit expensive...

As we're exposing `Git.git_repository_url` as a stateless helper function, this means that once per call, we're going to be reparsing/filtering the same repository url that never changes. Furthermore, when it's in git scheme (as per `url.rb`), we'll be throwing and handling an exception.

Ideally we would do this validation once (perhaps as part of the settings system?) and then reuse the known-to-be-valid value after that.

Reply:
Calling this parsing logic every time git_repository_url is called wasn't my intention at all. From my tests I see that caching in the module variable works and this code is called once per process, so it returns cached `@git_repository_url` every time. You can see here I have to remove these cached variables with some magic in tests in order to test different scenarios: https://github.com/DataDog/dd-trace-rb/pull/3424/files#diff-b489b0fca6f1f662bc61fa750ef1f1fbbe60e8fa7894e99b13606d1d05140ba0R8

So my intention is to write code that will be executed **at most once** per process. I chose for this the same approach that you implemented for ClassCount:

```ruby
module Datadog
  module Core
    module Environment
      # Retrieves number of classes from runtime
      module ClassCount
        def self.value
          ::ObjectSpace.count_objects[:T_CLASS]
        end

        def self.available?
          return @class_count_available if defined?(@class_count_available)

          @class_count_available =
            ::ObjectSpace.respond_to?(:count_objects) && ::ObjectSpace.count_objects.key?(:T_CLASS)
        end
      end
    end
  end
end
```

Please tell me if this does not work as intended and the ClassCount optimization was done with a different goal in mind.

Reply:
@ivoanjo please tell me if this approach would also work for profiling to reuse this info and if not, I am happy to make changes so that it works for every product

Reply:
> Git.git_repository_url as a stateless helper function 

I think this is not correct. Although naming with `util` suggest it is a stateless helper function, but the value is cached by its implementation. 

I think the performance impact should not be blocking this implementation.

Reply:
Ah, derp, you're completely right. It does work as you state :)

The one case it wouldn't ~exactly work like that is module inclusion. I've been trying to avoid the use of `module_function`, since I don't think it makes any sense at all that you can `include Core::Environment::Git` anywhere to directly access `@git_repository_url`. And that one would mean the cache would belong to every instance of whatever class included that module.

I actually got rid of `module_function` in the file you shared exactly for that reason.

---

(More as a generic comment, I think needing to `remove_instance_variable` in the tests is a code smell and in general I still kinda favour having state in objects, not class/modules BUT it's not like this pattern isn't in use in a lot of places in ddtrace, so I'm ok with saying "let's fix it another day).

Reply:
> @ivoanjo please tell me if this approach would also work for profiling to reuse this info and if not, I am happy to make changes so that it works for every product

Yeah, it works. If you want to take a stab at the profiling part (and I'd thank you), the right place to modify is: 
https://github.com/DataDog/dd-trace-rb/blob/master/lib/datadog/profiling/tag_builder.rb

Let me know / happy to help if you run into any troubles with it.

Reply:
Yeah, I will be happy to do it in TagBuilder too and I will also remove `module_function` too.

I hate using `remove_instance_variable` too, but I think here we have a tradeoff between needs to keep global state somewhere/test the code/make it performant.